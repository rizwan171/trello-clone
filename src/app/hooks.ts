import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
import type { RootState, AppDispatch } from "./store";

// The below useAppDispatch does not require a return type, hence the eslint disable comment
// See https://stackoverflow.com/questions/70711795/missing-return-type-on-function-useappdispatch for more info
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
