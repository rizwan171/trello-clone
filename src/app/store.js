import { configureStore } from "@reduxjs/toolkit";
import boardReducer from "../features/boardSlice.js";
import listsReducer from "../features/listsSlice.js";
import cardsReducer from "../features/cardsSlice.js";
import boardOptionsReducer from "../features/boardOptionSlice.js";
import currentSelectedCardReducer from "../features/currentSelectedCardSlice.js";
import currentSelectedListReducer from "../features/currentSelectedListSlice.js";
import listOptionsMenuPositionReducer from "../features/listOptionsMenuPositionSlice.js";
import tagsReducer from "../features/tagsSlice.js";
import selectedTagColourReducer from "../features/selectedTagColourSlice.js";
import createTagMenuDataReducer from "../features/createTagMenuDataSlice.js";
import modalActionMenusVisibilityReducer from "../features/modalActionMenusVisibilitySlice.js";

export const store = configureStore({
  reducer: {
    board: boardReducer,
    lists: listsReducer,
    cards: cardsReducer,
    boardOptions: boardOptionsReducer,
    currentSelectedCard: currentSelectedCardReducer,
    currentSelectedList: currentSelectedListReducer,
    listOptionsMenuPosition: listOptionsMenuPositionReducer,
    tags: tagsReducer,
    selectedTagColour: selectedTagColourReducer,
    createTagMenuData: createTagMenuDataReducer,
    modalActionMenusVisibility: modalActionMenusVisibilityReducer,
  },
});
