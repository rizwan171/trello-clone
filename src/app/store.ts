import { combineReducers, configureStore, PreloadedState } from "@reduxjs/toolkit";
import boardReducer from "../features/boardSlice";
import listsReducer from "../features/listsSlice";
import cardsReducer from "../features/cardsSlice";
import boardOptionsReducer from "../features/boardOptionSlice";
import currentSelectedCardReducer from "../features/currentSelectedCardSlice";
import currentSelectedListReducer from "../features/currentSelectedListSlice";
import listOptionsMenuPositionReducer from "../features/listOptionsMenuPositionSlice";
import tagsReducer from "../features/tagsSlice";
import selectedTagColourReducer from "../features/selectedTagColourSlice";
import createTagMenuDataReducer from "../features/createTagMenuDataSlice";
import modalActionMenusVisibilityReducer from "../features/modalActionMenusVisibilitySlice";

const rootReducer = combineReducers({
  board: boardReducer,
  lists: listsReducer,
  cards: cardsReducer,
  boardOptions: boardOptionsReducer,
  currentSelectedCard: currentSelectedCardReducer,
  currentSelectedList: currentSelectedListReducer,
  listOptionsMenuPosition: listOptionsMenuPositionReducer,
  tags: tagsReducer,
  selectedTagColour: selectedTagColourReducer,
  createTagMenuData: createTagMenuDataReducer,
  modalActionMenusVisibility: modalActionMenusVisibilityReducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = AppStore['dispatch']
export type AppStore = ReturnType<typeof setupStore>