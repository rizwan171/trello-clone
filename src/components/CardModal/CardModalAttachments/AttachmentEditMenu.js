import localforage from "localforage";
import React, { useState, useRef } from "react";
import { MdClose } from "react-icons/md";
import { useDispatch, useSelector } from "react-redux";
import { removeFileFromCard, updateFileInCard } from "../../../features/cardsSlice";
import { setCurrentSelectedCard } from "../../../features/currentSelectedCardSlice";
import { closeMenu } from "../../../features/modalActionMenusVisibilitySlice";

const AttachmentEditMenu = ({id, name}) => {
  const [fileName, setFileName] = useState(name);
  const card = useSelector((state) => state.currentSelectedCard.value);
  const dispatch = useDispatch();

  const handleEdit = () => {
    dispatch(updateFileInCard({ cardId: card.id, id, name: fileName}));
    const index = card.attachments.findIndex((item)=> item.fileId===id)
    let attachments = card.attachments.map(item=> {
        if (item.fileId === id) {
            return {fileId: id, date: item.date, name: fileName}
        }
        return item
    });
    if (index !== -1) {
    
        dispatch(setCurrentSelectedCard({...card, attachments}));    
    }

    handleClose();
  }

  const handleClose = () => {
    dispatch(closeMenu());
  };

  return (
    // TODO change mt-32 to something more robust
    <div className="fixed mt-16 w-80 ml-24 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4">
      <div className="relative text-center mb-2">
        <span className="text-sm block relative z-10">Edit attachment?</span>
        <MdClose onClick={handleClose} size={20} className="absolute right-0 top-0 z-20 cursor-pointer" />
      </div>
      <hr />
      <div className="flex flex-col mt-2 text-sm">
        <span className='font-medium pb-2'>Link name</span>

        <input
          type="text"
          autoFocus
          onChange={(e)=> setFileName(e.target.value.trim())}
          value = {fileName}
          className="w-full h-9 py-1 px-2 border-1 mb-2 border-gray-200 rounded-sm focus:outline-none focus:ring-2 focus:ring-trello-blue-100"
          placeholder="Change attachment name"
        />

        <button
          className="p-2 mt-2 w-32 bg-trello-blue-100 hover:bg-trello-blue-200 text-white font-semibold items-center text-sm shadow-sm rounded-sm"
          onClick={()=> handleEdit()}
        >
          Update
        </button>
      </div>
    </div>
  );
};

export default AttachmentEditMenu;
