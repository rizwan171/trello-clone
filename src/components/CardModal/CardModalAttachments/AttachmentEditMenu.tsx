import { useState } from "react";
import { MdClose } from "react-icons/md";
import { useAppDispatch, useAppSelector } from "../../../app/hooks";
import { updateFileInCard } from "../../../features/cardsSlice";
import { setCurrentSelectedCard } from "../../../features/currentSelectedCardSlice";
import { closeMenu } from "../../../features/modalActionMenusVisibilitySlice";
import { AttachmentEditMenuProps } from "../../../types/components/AttachmentEditMenuProps";

const AttachmentEditMenu = ({ fileId, fileName }: AttachmentEditMenuProps) => {
  const dispatch = useAppDispatch();
  const card = useAppSelector((state) => state.currentSelectedCard.value);
  const [editableFileName, setEditableFileName] = useState(fileName);

  const handleEdit = () => {
    if (!card) return;

    dispatch(updateFileInCard({ cardId: card.id, fileId, fileName: editableFileName }));
    const index = card.attachments.findIndex((item) => item.id === fileId);
    const attachments = card.attachments.map((item) => {
      if (item.id === fileId) {
        return { id: fileId, date: item.date, name: editableFileName };
      }
      return item;
    });
    if (index !== -1) {
      dispatch(setCurrentSelectedCard({ ...card, attachments: [...attachments] }));
    }

    handleClose();
  };

  const handleClose = () => {
    dispatch(closeMenu());
  };

  return (
    // TODO change mt-32 to something more robust
    <div className="fixed mt-16 w-80 ml-24 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4">
      <div className="relative text-center mb-2">
        <span className="text-sm block relative z-10">Edit attachment?</span>
        <MdClose onClick={handleClose} size={20} className="absolute right-0 top-0 z-20 cursor-pointer" />
      </div>
      <hr />
      <div className="flex flex-col mt-2 text-sm">
        <span className="font-medium pb-2">Link name</span>

        <input
          type="text"
          autoFocus
          onChange={(e) => setEditableFileName(e.target.value.trim())}
          value={editableFileName}
          className="w-full h-9 py-1 px-2 border-1 mb-2 border-gray-200 rounded-sm focus:outline-none focus:ring-2 focus:ring-trello-blue-100"
          placeholder="Change attachment name"
        />

        <button
          className="p-2 mt-2 w-32 bg-trello-blue-100 hover:bg-trello-blue-200 text-white font-semibold items-center text-sm shadow-sm rounded-sm"
          onClick={() => handleEdit()}
        >
          Update
        </button>
      </div>
    </div>
  );
};

export default AttachmentEditMenu;
