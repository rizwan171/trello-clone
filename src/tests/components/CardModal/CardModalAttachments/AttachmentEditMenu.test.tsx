import { fireEvent, screen } from "@testing-library/react";
import { RootState } from "../../../../app/store";
import AttachmentEditMenu from "../../../../components/CardModal/CardModalAttachments/AttachmentEditMenu";
import { renderWithProviders } from "../../../utils/renderUtils";
import { generateModalActionsVisibilityState } from "../../../utils/stateUtils";

describe("AttatchmentEditMenu", () => {
  let initialState: Partial<RootState>;
  let fileId: string;
  let fileName: string;

  beforeEach(() => {
    fileId = "fileId";
    fileName = "Test File Name";

    const card = {
      id: "1",
      title: "Card 1",
      description: "Description",
      listId: "1",
      tags: [],
      attachments: [{ id: fileId, name: fileName, date: new Date().toString() }],
    };

    const modalActionsVisibilityState = generateModalActionsVisibilityState();

    initialState = {
      cards: {
        value: [{ ...card }],
      },
      currentSelectedCard: {
        value: { ...card },
      },
      modalActionMenusVisibility: {
        value: {
          ...modalActionsVisibilityState.value,
          attachmentEditMenuOpen: {
            id: 1,
            status: true,
          },
        },
      },
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render successfully", () => {
    const view = renderWithProviders(<AttachmentEditMenu fileId={fileId} fileName={fileName} />, {
      preloadedState: initialState,
    });
    expect(view.asFragment()).toMatchSnapshot();
  });

  it("should edit the attachment name", () => {
    const { store } = renderWithProviders(<AttachmentEditMenu fileId={fileId} fileName={fileName} />, {
      preloadedState: initialState,
    });
    expect(store.getState().cards.value[0].attachments).toHaveLength(1);
    expect(store.getState().cards.value[0].attachments[0].name).toBe(fileName);
    expect(store.getState().currentSelectedCard.value?.attachments).toHaveLength(1);
    expect(store.getState().currentSelectedCard.value?.attachments[0].name).toBe(fileName);
    expect(store.getState().modalActionMenusVisibility.value.attachmentEditMenuOpen.status).toBeTruthy();

    const editInput = screen.getByTestId("attachment-edit-menu-input") as HTMLInputElement;
    expect(editInput).toBeInTheDocument();

    const updatedNameString = "Updated Name";
    fireEvent.change(editInput, { target: { value: updatedNameString } });

    const updateButton = screen.getByText("Update") as HTMLButtonElement;
    expect(updateButton).toBeInTheDocument();

    fireEvent.click(updateButton);

    expect(store.getState().cards.value[0].attachments[0].name).toBe(updatedNameString);
    expect(store.getState().currentSelectedCard.value?.attachments[0].name).toBe(updatedNameString);
    expect(store.getState().modalActionMenusVisibility.value.attachmentEditMenuOpen.id).toBe(0);
    expect(store.getState().modalActionMenusVisibility.value.attachmentEditMenuOpen.status).toBeFalsy();
  });

  it("should close the edit attachment menu", () => {
    const { store } = renderWithProviders(<AttachmentEditMenu fileId={fileId} fileName={fileName} />, {
      preloadedState: initialState,
    });
    expect(store.getState().modalActionMenusVisibility.value.attachmentEditMenuOpen.status).toBeTruthy();

    const closeIcon = screen.getByTestId("attachment-edit-menu-close");
    expect(closeIcon).toBeInTheDocument();

    fireEvent.click(closeIcon);

    expect(store.getState().modalActionMenusVisibility.value.attachmentEditMenuOpen.id).toBe(0);
    expect(store.getState().modalActionMenusVisibility.value.attachmentEditMenuOpen.status).toBeFalsy();
  });
});
