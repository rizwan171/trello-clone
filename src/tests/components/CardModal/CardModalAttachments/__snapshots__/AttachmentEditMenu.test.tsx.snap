// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttatchmentEditMenu should render successfully 1`] = `
<DocumentFragment>
  <div
    class="fixed mt-16 w-80 ml-24 text-gray-700 bg-white rounded-ibsm shadow-2xl p-4"
  >
    <div
      class="relative text-center mb-2"
    >
      <span
        class="text-sm block relative z-10"
      >
        Edit attachment?
      </span>
      <svg
        class="absolute right-0 top-0 z-20 cursor-pointer"
        data-testid="attachment-edit-menu-close"
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <hr />
    <div
      class="flex flex-col mt-2 text-sm"
    >
      <span
        class="font-medium pb-2"
      >
        Link name
      </span>
      <input
        class="w-full h-9 py-1 px-2 border-1 mb-2 border-gray-200 rounded-sm focus:outline-none focus:ring-2 focus:ring-trello-blue-100"
        data-testid="attachment-edit-menu-input"
        placeholder="Change attachment name"
        type="text"
        value="Test File Name"
      />
      <button
        class="p-2 mt-2 w-32 bg-trello-blue-100 hover:bg-trello-blue-200 text-white font-semibold items-center text-sm shadow-sm rounded-sm"
      >
        Update
      </button>
    </div>
  </div>
</DocumentFragment>
`;
